docs: using AsyncAPI with Kafka (#2252)
asyncapi/website: pages/docs/tutorials/kafka/index.md

---

> title: Create AsyncAPI Document for applications interacting with Kafka

Reviewer: this is a specific tutorial that describes application that is consumer of kafka messages, as it has receive action

```suggestion
title: Create AsyncAPI Document for applications consuming from Kafka
```

wdyt? just a suggestion, can be generic as it is if you think it is better

description: Explore configuring AsyncAPI document for Kafka messages.
weight: 20
---

## Introduction

> The tutorial will walk you through the process of creating an schema document for Kafka Messages using AsyncAPI. Additionally, you will also learn about event-driven architecture, pub/sub model, and brokers in relation to Kafka. 

Reviewer 2: ```suggestion
The tutorial will walk you through the process of creating a schema document for Kafka Messages using AsyncAPI. Additionally, you will also learn about event-driven architecture, pub/sub model, and brokers in relation to Kafka. 
```

  
Let’s assume, you have a service that publishes data to a Kafka topic whenever a new user signs up into the application.

> Basically, you want to build a system that can track the users who have signed up for your application, the entire flow of the application would be:

Reviewer 2: ```suggestion
Basically, if you want to build a system that can track the users who have signed up for your application, the entire flow of the application would be:
```

```mermaid

> sequenceDiagram

Reviewer: great diagram 👏🏼 
just please rename Message Broker to Kafka since it is Kafka related

    participant Your Application (Consumer)
    participant Message Broker
    participant Service (Producer)
    participant User

>     Your Application (Consumer)->> Message Broker: Connects to "userSignedUp" topic

Reviewer: ```suggestion
    Your Application (Consumer)->> Message Broker: Subscribes to "userSignedUp" topic
```

    User-->> Service (Producer): Signs Up
    Service (Producer)->> Message Broker: Publishes Events

>     Message Broker->> Your Application (Consumer): Sends "user_signup" payload

Reviewer: ```suggestion
    Message Broker->> Your Application (Consumer): Sends "user_signup" message
```

```    

## Background context

[Event-driven architecture (EDA)](/docs/tutorials/getting-started/event-driven-architectures) is a design pattern that revolves around the production, detection, and reaction to events over time. It consists of three main components: a message broker, event publishers, and subscribers, which together serve as the backbone for event exchange within different services. 

[Message brokers](/docs/tutorials/getting-started/event-driven-architectures#message-broker) facilitate asynchronous communications between services, meaning that the sending service doesn't have to wait for the receiving service's response. This allows multiple services to talk to each other directly, even if they are written in different languages. A great example of such a message broker is Apache Kafka. [Apache Kafka](https://kafka.apache.org/) is a distributed system that is used by thousands of companies for their event-driven system.

> [Publish/Subscribe (Pub/Sub)](/docs/tutorials/getting-started/event-driven-architectures#publishersubscriber) is a typical model in EDAs, providing flexible coupling between publishers and subscribers. In this model, the publishers of event notifications act as publishers and the consumers act as corresponding subscribers.

Reviewer 2: ```
the publishers of event notifications act as publishers
```
Can this be phrased differently? It's a little confusing.

> [Publish/Subscribe (Pub/Sub)](/docs/tutorials/getting-started/event-driven-architectures#publishersubscriber) is a typical model in EDAs, providing flexible coupling between publishers and subscribers. In this model, those generating event notifications act as publishers, while those receiving the notifications act as the corresponding subscribers.

Reviewer: ```suggestion
[Publisher/Subscriber (Pub/Sub)](/docs/tutorials/getting-started/event-driven-architectures#publishersubscriber) is a typical model in EDAs, providing flexible coupling between publishers and subscribers. In this model, those that `send` event notifications act as publishers, while those that `receive` the notifications act as the corresponding subscribers.
```

`Publish/Subscribe` are cursed words here 😄 
also made sure we somehow refer to send/receive that is new

## Creating AsyncAPI document for Kafka Messages  

In this section, you’ll create an AsyncAPI document to describe the `UserSignUp` API. The same document can be later used to generate code and documentation as per requirement.

Let's begin by analysing the document. As we move forward in the tutorial, we'll break down each segment uncovering its purpose and functionality.

```
asyncapi: 3.0.0
info:
  title: User Signup API
  version: 1.0.0
  description: The API notifies you whenever a new user signs up in the application.

> servers:

Reviewer: since this is kafka related tutorial, please add more servers and explain as Dale does it in https://www.youtube.com/watch?v=Ni5tCY9r0TY. It is old tutorial but except of rename from `url` to `host`, servers work the same way in v3

        properties:
          user-id:
            type: integer
            description: This property describes the id of the user
          user-email:
            type: string
            description: This property describes the email of the user
```

Let's now break it down into pieces:

### Define AsyncAPI version, API information, and server

Initially, you need to describe your application, including the AsyncAPI version, the info about the document, and the server your application is based upon.

```
asyncapi: 3.0.0
info:
  title: User Signup API
  version: 1.0.0
  description: The API notifies you whenever a new user signs up in the application.
servers:
  kafkaServer:
    host: test.mykafkacluster.org:8092
    description: Kafka Server
    protocol: kafka
```

In the above snippet:

- The `asyncapi` field indicates that you are using AsyncAPI version 3.0.0.
  
- The `info` field provides information about the API. Here the the APIs `title`, `version`, and `description` are being defined.
  
- The `server` field specifies the details of the server, including the `host`, `description`, and the `protocol` that is being used i.e. Kafka.

### Define channels and operations

> Next, let's move on to the `channels` and `operations` section. The channel addresses are the topics in Kafka, they are the routes to which your API will be sending/receiving. The `operations` property is used to describe how your application interacts with the channels.

Reviewer 2: ```suggestion
Next, let's move on to the `channels` and `operations` section. The channel addresses are the topics in Kafka, they are the routes to which your API will be sending/receiving messages. The `operations` property is used to describe how your application interacts with the channels.
```

```
operations:
  onUserSignedUp:
    action: receive
    channel:
      $ref: '#/channels/userSignedUp'

channels:
  userSignedUp:
    description: This channel contains a message per each user who signs up in our application.
    address: user_signedup
    messages:
      userSignedUp:
        $ref: '#/components/messages/userSignedUp'
```

In the above snippet:

- The `onUserSignedUp` object specifies the nature of the operation. The `action` property indicates that our application will be receiving the information. The `channel` property points to the channel where the operation occurs.
  
- The `userSignedUp` object inside `channels` describes the Kafka topic where our application will be receiving the information and the associated message definition. The `address` field represents the actual name of the Kafka topic. The `messages` field describes the expected messages in that topic. 

### Define messages and schemas

> Finally, you'll define the messages and their payload. The payload defines how the event would look line that will be sent from the channel.

Reviewer 2: ```suggestion
Finally, you'll define the messages and their payload. The payload defines how the event would look like when it's sent from the channel.
```
Fixed a typo

---
title: Create AsyncAPI Document for applications consuming from Kafka
description: Explore configuring AsyncAPI document for Kafka messages.
weight: 20
---

> 

Reviewer: just an observation, something you could consult with other writers

tutorials set called Getting Started have a dedicated short `Introduction` document that is placed in `index.md` as a main document for getting started, an intro (referring to it as you also place first tutorial in `index.md`)

since there are going to be 4 kafka tutorials, maybe it would make sense to have a separate introduction to these? with generic background and short explanation about the flow, what given tutorials is about and the suggested order of reading?

anyway, just an idea you could explore and discuss with others

Author: Sounds good, I'll discuss it with others 👍

Reviewer 3: Not for this case because this is not a Quickstart. We wrote a 4 part tutorial series, but they do not all share the same introduction. 

We have been using the "Introduction" in all tutorials as the appropriate place to introduce your readers to the goal of that specific tutorial, so I would prefer we retain consistency. 😄

## Introduction

The tutorial will walk you through the process of creating a schema document for Kafka Messages using AsyncAPI. Additionally, you will also learn about event-driven architecture, pub/sub model, and brokers in relation to Kafka. 
  
Let’s assume, you have a service that publishes data to a Kafka topic whenever a new user signs up into the application.

Basically, if you want to build a system that can track the users who have signed up for your application, the entire flow of the application would be:

```mermaid
sequenceDiagram
    participant Your Application (Consumer)
    participant Kafka
    participant Service (Producer)
    participant User
    Your Application (Consumer)->> Kafka: Subscribes to "userSignedUp" topic
    User-->> Service (Producer): Signs Up
    Service (Producer)->> Kafka: Publishes Events
    Kafka->> Your Application (Consumer): Sends "user_signup" message
```    

> ## Background context

Reviewer: shouldn't there be some intro to Kafka?

[Event-driven architecture (EDA)](/docs/tutorials/getting-started/event-driven-architectures) is a design pattern that revolves around the production, detection, and reaction to events over time. It consists of three main components: a message broker, event publishers, and subscribers, which together serve as the backbone for event exchange within different services. 

> [Message brokers](/docs/tutorials/getting-started/event-driven-architectures#message-broker) facilitate asynchronous communications between services, meaning that the sending service doesn't have to wait for the receiving service's response. This allows multiple services to talk to each other directly, even if they are written in different languages. A great example of such a message broker is Apache Kafka. [Apache Kafka](https://kafka.apache.org/) is a distributed system that is used by thousands of companies for their event-driven system.

Author: > shouldn't there be some intro to Kafka?

@derberg I have already defined Kafka while defining Message Broker, should I add more introduction?

Reviewer: might be ok, leaving it up to you and other writers

with my experience I might be biased

Author: Personally I think this much content is enough, but I'll try checking with others

Reviewer 3: I actually liked how you seamlessly defined message brokers and then used that as the starting point to define Kafka next.

That said, I also agree with Lukasz that it initially feels as if we "forgot" to also introduce/define Kafka. I would vote to move your Kafka sentences into their own paragraph, so it doesn't get lost in the message broker definition.  I would also consider if you can add a sentence to describe what is a "distributed system" because it's not clear how that relates to the term "message brokers". Lemme know if that makes sense 😄

Author: Yeah, it makes sense. I'll make the changes ASAP

  description: The API notifies you whenever a new user signs up in the application.
servers:
  centralKafkaServer:
    host: central.mykafkacluster.org:8092
    description: Kafka broker running in central warehouse
    protocol: kafka
  westKafkaServer:
    host: west.mykafkacluster.org:8092
    description: Kafka broker running in west warehouse
    protocol: kafka
  eastKafkaServer:
    host: east.mykafkacluster.org:8092
    description: Kafka broker running in east warehouse
    protocol: kafka
```

In the above snippet:

- The `asyncapi` field indicates that you are using AsyncAPI version 3.0.0.
  
- The `info` field provides information about the API. Here the the APIs `title`, `version`, and `description` are being defined.
  
- The `server` field specifies the details of the server, including the `host`, `description`, and the `protocol` that is being used i.e., Kafka. This section allows you to define multiple clusters or servers, as shown in the provided code block.

<Remember>
If an application has multiple servers hidden from the user behind a load balancer, then only the address of the load balancer needs to be included in the AsyncAPI document. 
</Remember>

> ### Define channels and operations

Reviewer: I suggest you split it into 2 sections, separately channels and operations as they are very important and good to have it clearly separated in doc structure too

```

In the above snippet:

- The `onUserSignedUp` object specifies the nature of the operation. The `action` property indicates that our application will be receiving the information. The `channel` property points to the channel where the operation occurs.
  
- The `userSignedUp` object inside `channels` describes the Kafka topic where our application will be receiving the information and the associated message definition. The `address` field represents the actual name of the Kafka topic. The `messages` field describes the expected messages in that topic. 

### Define messages and schemas

Finally, you'll define the messages and their payload. The payload defines how the event would look like when it's sent from the channel.

```
components:
  messages:
    userSignedUp:
      payload:
        type: object
        properties:
          user-id:
            type: integer
            description: This property describes the id of the user
          user-email:
            type: string
            description: This property describes the id of the user
```

In the above snippet:

- The `userSignedUp` message is defined which describes the payload (content) of the message.
  

> - The `payload` property defines the content of the message using [JSON Schema](https://json-schema.org/). It means that your message payload should contain a `user-id` which is an integer and a `user-email` property which is a string property.

Reviewer: you use AsyncAPI schema, not JSON schema here, AsyncAPI schema is default

if you would use JSON schema, you would have to do it like in case of Avro schema, with `schemaFormat`

---

> title: Create AsyncAPI Document for applications consuming from Kafka

Reviewer 3: ```suggestion
title: Create AsyncAPI document for applications consuming from Kafka
```

> description: Explore configuring AsyncAPI document for Kafka messages.

Reviewer 3: ```suggestion
description: A tutorial teaching how to configure an AsyncAPI document for Kafka messages.
```

weight: 20
---

## Introduction

> The tutorial will walk you through the process of creating a schema document for Kafka Messages using AsyncAPI. Additionally, you will also learn about event-driven architecture, pub/sub model, and brokers in relation to Kafka. 

Reviewer 3: ```suggestion
In this tutorial, you will create a schema document for Kafka Messages using AsyncAPI. You will also learn about Event-Driven Architecture, the pub/sub model, and brokers concerning Kafka. 
```

> Let’s assume, you have a service that publishes data to a Kafka topic whenever a new user signs up into the application.

Reviewer 3: ```suggestion
Suppose you have a service publishing data to a Kafka topic whenever a new user signs up on your application.
```

> Basically, if you want to build a system that can track the users who have signed up for your application, the entire flow of the application would be:

Reviewer 3: ```suggestion
You want to build a system that tracks users who have signed up for your application. As such, your application will have the following flow:
```

```mermaid
sequenceDiagram
    participant Your Application (Consumer)
    participant Kafka
    participant Service (Producer)
    participant User
    Your Application (Consumer)->> Kafka: Subscribes to "userSignedUp" topic
    User-->> Service (Producer): Signs Up
    Service (Producer)->> Kafka: Publishes Events
    Kafka->> Your Application (Consumer): Sends "user_signup" message
```    

## Background context

> [Event-driven architecture (EDA)](/docs/tutorials/getting-started/event-driven-architectures) is a design pattern that revolves around the production, detection, and reaction to events over time. It consists of three main components: a message broker, event publishers, and subscribers, which together serve as the backbone for event exchange within different services. 

Reviewer 3: ```suggestion
[Event-Driven Architecture (EDA)](/docs/tutorials/getting-started/event-driven-architectures) is a design pattern that revolves around the production, detection, and reaction to events over time. It consists of three main components: a message broker, event publishers, and subscribers, which together serve as the backbone for event exchange within different services. 
```

> [Message brokers](/docs/tutorials/getting-started/event-driven-architectures#message-broker) facilitate asynchronous communications between services, meaning that the sending service doesn't have to wait for the receiving service's response. This allows multiple services to talk to each other directly, even if they are written in different languages. 

Reviewer 3: ```suggestion
[Message brokers](/docs/tutorials/getting-started/event-driven-architectures#message-broker) facilitate asynchronous communications between services, meaning the sending service doesn't have to wait for the receiving service's response. Multiple services can talk to each other directly, even if they are written in different languages. 
```

> A great example of such a message broker is Apache Kafka. [Apache Kafka](https://kafka.apache.org/) is a distributed event streaming platform that is used by thousands of companies for their event-driven system. Here, a distributed event streaming platform refers to a system that operates on multiple nodes, providing high availability and fault tolerance.

Reviewer 3: ```suggestion
A great example of such a message broker is Apache Kafka. [Apache Kafka](https://kafka.apache.org/) is a distributed event streaming platform thousands of companies use for their event-driven system. Here, a distributed event streaming platform is a system that operates on multiple nodes, providing high availability and fault tolerance.
```

> [Publisher/Subscriber (Pub/Sub)](/docs/tutorials/getting-started/event-driven-architectures#publishersubscriber) is a typical model in EDAs, providing flexible coupling between publishers and subscribers. In this model, those that `send` event notifications act as publishers, while those that `receive` the notifications act as the corresponding subscribers.

Reviewer 3: ```suggestion
[Publisher/Subscriber (Pub/Sub)](/docs/tutorials/getting-started/event-driven-architectures#publishersubscriber) is a typical model in EDAs, providing flexible coupling between publishers and subscribers. In this model, those who `send` event notifications act as publishers, while those who `receive` the notifications act as the corresponding subscribers.
```

> ## Creating AsyncAPI document for Kafka Messages  

Reviewer 3: ```suggestion
## Create AsyncAPI document for Kafka messages  
```

In this section, you’ll create an AsyncAPI document to describe the `UserSignUp` API. The same document can be later used to generate code and documentation as per requirement.

> Let's begin by analysing the document. As you move forward in the tutorial, we'll break down each segment uncovering its purpose and functionality.

Reviewer 3: ```suggestion
Let's begin by analyzing the document. As you progress in the tutorial, we'll break down each segment, uncovering its purpose and functionality.
```

  description: The API notifies you whenever a new user signs up in the application.

servers:
  centralKafkaServer:
    host: central.mykafkacluster.org:8092
    description: Kafka broker running in a central warehouse
    protocol: kafka
  westKafkaServer:
    host: west.mykafkacluster.org:8092
    description: Kafka broker running in the west warehouse
    protocol: kafka
  eastKafkaServer:
    host: east.mykafkacluster.org:8092
    description: Kafka broker running in the east warehouse
    protocol: kafka

operations:
  onUserSignedUp:
    action: receive
    channel:
      $ref: '#/channels/userSignedUp'

channels:
  userSignedUp:
    description: This channel contains a message per each user who signs up in our application.
    address: user_signedup
    messages:
      userSignedUp:
        $ref: '#/components/messages/userSignedUp'

components:
  messages:
    userSignedUp:
      payload:
        type: object
        properties:
          userId:
            type: integer
            description: This property describes the id of the user
          userEmail:
            type: string
            description: This property describes the email of the user
```

> Let's now break it down into pieces:

Reviewer 3: ```suggestion
Let's now break it down into pieces.
```

### Define AsyncAPI version, API information, and server

> Initially, you need to describe your application, including the AsyncAPI version, the info about the document, and the server your application is based upon.

Reviewer 3: ```suggestion
First, you must describe your application, including the AsyncAPI version, the info about the document, and the server your application is based on.
```

```
asyncapi: 3.0.0
info:
  title: User Signup API
  version: 1.0.0
  description: The API notifies you whenever a new user signs up in the application.
servers:
  centralKafkaServer:
    host: central.mykafkacluster.org:8092
    description: Kafka broker running in a central warehouse
    protocol: kafka
  westKafkaServer:
    host: west.mykafkacluster.org:8092
    description: Kafka broker running in the west warehouse
    protocol: kafka
  eastKafkaServer:
    host: east.mykafkacluster.org:8092
    description: Kafka broker running in the east warehouse
    protocol: kafka
```

In the above snippet:

> - The `asyncapi` field indicates that you are using AsyncAPI version 3.0.0.

Reviewer 3: ```suggestion
- The `asyncapi` field indicates that you use AsyncAPI version 3.0.0.
```

  
- The `info` field provides information about the API. Here the the APIs `title`, `version`, and `description` are being defined.
  

> - The `server` field specifies the details of the server, including the `host`, `description`, and the `protocol` that is being used i.e., Kafka. This section allows you to define multiple clusters or servers, as shown in the provided code block.

Reviewer 3: ```suggestion
- The `server` field specifies the details of the server, including the `host`, `description`, and the `protocol` that is being used (i.e., Kafka). This section allows you to define multiple clusters or servers, as shown in the provided code block.
```

<Remember>

> If an application has multiple servers hidden from the user behind a load balancer, then only the address of the load balancer needs to be included in the AsyncAPI document. 

Reviewer 3: ```suggestion
If an application has multiple servers hidden from the user behind a load balancer, then only the address of the load balancer must be included in the AsyncAPI document. 
```

</Remember>

### Define channels

> Next, let's move on to the `channels` section. The channel addresses are the topics in Kafka, they are the routes to which your API will be sending/receiving messages. 

Reviewer 3: ```suggestion
Next, let's move on to the `channels` section. The channel addresses are the topics in Kafka; they are the routes to which your API will send/receive messages. 
```

```
channels:
  userSignedUp:
    description: This channel contains a message per each user who signs up in our application.
    address: user_signedup
    messages:
      userSignedUp:
        $ref: '#/components/messages/userSignedUp'
```

> In the above snippet:

Reviewer 3: ```suggestion
In the above code snippet:
```

> - The `userSignedUp` object inside `channels` describes the Kafka topic where the application will be receiving the information and the associated message definition. 

Reviewer 3: ```suggestion
- The `userSignedUp` object inside `channels` describes the Kafka topic where the application will receive the information and the associated message definition. 
```

- The `address` field represents the actual name of the Kafka topic and the `messages` field describes the expected messages in that topic.

### Define operations

Now, let's move on to the `operations` section. The `operations` property is used to describe how your application interacts with the defined channels.

```
operations:
  onUserSignedUp:
    action: receive
    channel:
      $ref: '#/channels/userSignedUp'
```

> In the above snippet:

Reviewer 3: ```suggestion
In the above code snippet:
```

> - The `onUserSignedUp` object specifies the nature of the operation, that occurs when user sign-ups. 
> - The `action` property indicates that the application will be receiving the information.
> - The `channel` property points to the channel where the operation occurs.

Reviewer 3: ```suggestion
- The `onUserSignedUp` object specifies the operation's nature, which occurs when the user signs up. 
- The `action` property indicates that the application will receive the information.
- The `channel` property points to the channel where the operation occurs.
```

### Define messages and schemas

Finally, you'll define the messages and their payload. The payload defines how the event would look like when it's sent from the channel.

```
components:
  messages:
    userSignedUp:
      payload:
        type: object
        properties:
          userId:
            type: integer
            description: This property describes the ID of the user
          userEmail:
            type: string
            description: This property describes the ID of the user
```

> In the above snippet:

Reviewer 3: ```suggestion
In the above code snippet:
```

- The `userSignedUp` message is defined which describes the payload (content) of the message.
  

> - The `payload` property defines the content of the message. It means that your message payload should contain a `userId` which is an integer and a `userEmail` property which is a string property.

Reviewer 3: ```suggestion
- The `payload` property defines the message content. Your message payload should contain a `userId`, an integer, a `userEmail` property, and a string property.
```

## Summary

> The ability to generate an AsyncAPI document for Kafka is now in your toolkit. You generated an AsyncAPI document that defines the structure of the Kafka messages in a machine-readable format which makes it easier to maintain event-driven architecture. Try adding your own business logic and playing around with it.

Reviewer 3: ```suggestion
The ability to generate an AsyncAPI document for Kafka is now in your toolkit. You generated an AsyncAPI document that defines the structure of the Kafka messages in a machine-readable format, which makes it easier to maintain event-driven architecture. Try adding your business logic and playing around with it.
```

## Next steps

> Now that you know how to write an AsyncAPI document for Kafka messages using the default schema, let's proceed to learn how to compose an [AsyncAPI document using Avro Schema instead](/docs/tutorials/kafka/configure-kafka-avro). 

Reviewer 3: ```suggestion
Now that you know how to write an AsyncAPI document for Kafka messages using the default schema, let's learn how to create an [AsyncAPI document using Avro Schema instead](/docs/tutorials/kafka/configure-kafka-avro). 
```
